### PR TITLE
[feature](Nereids): one side eager aggregation

### PR DESCRIPTION
**Pull request by jackwener**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java [hunk 1]
```diff
@@ -275,6 +277,10 @@ public class Rewriter extends AbstractBatchJobExecutor {
                             new PushDownMinMaxThroughJoin(),
                             new PushDownCountThroughJoin()
                     ),
+                    topDown(
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java:280]: min/max just can push one side

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownCountThroughJoinOneSide.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+        // top Sum agg
+        // count(slot) -> sum( count(slot) as cnt )
+        List<NamedExpression> newOutputExprs = new ArrayList<>();
+        for (NamedExpression ne : agg.getOutputExpressions()) {
+            if (ne instanceof Alias && ((Alias) ne).child() instanceof Count) {
+                Count oldTopCnt = (Count) ((Alias) ne).child();
+
+                Slot slot = (Slot) oldTopCnt.child(0);
+                if (leftCntSlotToOutput.containsKey(slot)) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownCountThroughJoinOneSide.java:201]: would we do some pre-processing ahead, such as judge which side will push down, and at this point, we just do the transformation without the additional 'if-else'?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownSumThroughJoinOneSide.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+ * </pre>
+ */
+public class PushDownSumThroughJoinOneSide implements RewriteRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                logicalAggregate(innerLogicalJoin())
+                        .when(agg -> agg.child().getOtherJoinConjuncts().isEmpty())
+                        .whenNot(agg -> agg.child().children().stream().anyMatch(p -> p instanceof LogicalAggregate))
+                        .when(agg -> {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownSumThroughJoinOneSide.java:70]: some as the above comments, do 'one side only' judgment ahead here?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownSumThroughJoinOneSide.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+        // top Sum agg
+        // replace sum(x) -> sum(sum#)
+        List<NamedExpression> newOutputExprs = new ArrayList<>();
+        for (NamedExpression ne : agg.getOutputExpressions()) {
+            if (ne instanceof Alias && ((Alias) ne).child() instanceof Sum) {
+                Sum oldTopSum = (Sum) ((Alias) ne).child();
+
+                Slot slot = (Slot) oldTopSum.child(0);
+                if (leftSumSlotToOutput.containsKey(slot)) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownSumThroughJoinOneSide.java:194]: some as the above comments, do the definite side push transformation where without 'if-else' by moving some pre-processing at former logic?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownCountThroughJoinOneSide.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+ * Notice: rule can't optimize condition that groupby is empty when Count(*) exists.
+ */
+public class PushDownCountThroughJoinOneSide implements RewriteRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                logicalAggregate(innerLogicalJoin())
+                        .when(agg -> agg.child().getOtherJoinConjuncts().isEmpty())
+                        .whenNot(agg -> agg.child().children().stream().anyMatch(p -> p instanceof LogicalAggregate))
+                        .when(agg -> agg.getGroupByExpressions().stream().allMatch(e -> e instanceof Slot))
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownCountThroughJoinOneSide.java:75]: Not a big impact, change it together with the next PR

**Review by xzj7019 — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Comment by wuyanxing** [comment]
why a rbo not a cbo path? @jackwener

**Pull request merged**